File: src/zoho_cli/auth/commands.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated

import cappa

from zoho_cli.auth.config import (
    CONFIG_FILE,
    TOKENS_FILE,
    resolve_auth,
)
from zoho_cli.auth.device_flow import device_flow_login
from zoho_cli.auth.self_client import self_client_exchange
from zoho_cli.auth.token import refresh_access_token
from zoho_cli.errors import AuthError, err
from zoho_cli.output import output


@cappa.command(name="login", help="Authenticate via device flow OAuth")
@dataclass
class Login:
    client_id: Annotated[
        str,
        cappa.Arg(long="--client-id", help="Zoho OAuth client ID"),
    ]
    client_secret: Annotated[
        str,
        cappa.Arg(long="--client-secret", help="Zoho OAuth client secret"),
    ]
    dc: Annotated[
        str,
        cappa.Arg(
            long="--dc",
            default="com",
            help="Data center (com, eu, in, com.au, jp, ca, sa, uk, com.cn)",
        ),
    ] = "com"
    scopes: Annotated[
        str | None,
        cappa.Arg(
            long="--scopes", default=None, help="Comma-separated OAuth scopes (defaults to all)"
        ),
    ] = None

    def __call__(self) -> None:
        device_flow_login(
            client_id=self.client_id,
            client_secret=self.client_secret,
            dc=self.dc,
            scopes=self.scopes,
        )


@cappa.command(name="self-client", help="Authenticate via self-client code exchange")
@dataclass
class SelfClient:
    code: Annotated[
        str,
        cappa.Arg(long="--code", help="Self-client authorization code from Zoho API Console"),
    ]
    client_id: Annotated[
        str,
        cappa.Arg(long="--client-id", help="Zoho OAuth client ID"),
    ]
    client_secret: Annotated[
        str,
        cappa.Arg(long="--client-secret", help="Zoho OAuth client secret"),
    ]
    dc: Annotated[
        str,
        cappa.Arg(long="--dc", default="com", help="Data center"),
    ] = "com"
    server: Annotated[
        str | None,
        cappa.Arg(long="--server", default=None, help="Accounts server URL override"),
    ] = None

    def __call__(self) -> None:
        self_client_exchange(
            client_id=self.client_id,
            client_secret=self.client_secret,
            code=self.code,
            dc=self.dc,
            accounts_server=self.server,
        )


@cappa.command(name="status", help="Show current authentication status")
@dataclass
class Status:
    def __call__(self) -> None:
        try:
            config = resolve_auth()
        except AuthError:
            err("Not authenticated. Run `zoho auth login` to authenticate.")
            raise SystemExit(2) from None

        info = {
            "authenticated": True,
            "source": config.source,
            "dc": config.dc,
            "accounts_url": config.accounts_url,
            "token_valid": config.token_valid,
        }
        output(info)


@cappa.command(name="refresh", help="Force refresh the access token")
@dataclass
class Refresh:
    def __call__(self) -> None:
        config = resolve_auth()
        token = refresh_access_token(config)
        err("Access token refreshed successfully.")
        output({"access_token": token[:20] + "...", "status": "refreshed"})


@cappa.command(name="logout", help="Clear stored authentication tokens")
@dataclass
class Logout:
    def __call__(self) -> None:
        removed = False
        if TOKENS_FILE.exists():
            TOKENS_FILE.unlink()
            removed = True
        if CONFIG_FILE.exists():
            CONFIG_FILE.unlink()
            removed = True
        if removed:
            err("Logged out. Stored credentials removed.")
        else:
            err("No stored credentials found.")


@cappa.command(name="auth", help="Authentication management")
@dataclass
class Auth:
    subcommand: cappa.Subcommands[Login | SelfClient | Status | Refresh | Logout]

File: internal/auth/commands.go
package auth

import (
	"context"
	"fmt"
	"os"

	"github.com/omin8tor/zoho-cli/internal"
	"github.com/omin8tor/zoho-cli/internal/output"
	"github.com/urfave/cli/v3"
)

func Commands() *cli.Command {
	return &cli.Command{
		Name:  "auth",
		Usage: "Authentication management",
		Commands: []*cli.Command{
			{
				Name:  "login",
				Usage: "Authenticate via device flow OAuth",
				Flags: []cli.Flag{
					&cli.StringFlag{Name: "client-id", Required: true, Usage: "Zoho OAuth client ID"},
					&cli.StringFlag{Name: "client-secret", Required: true, Usage: "Zoho OAuth client secret"},
					&cli.StringFlag{Name: "dc", Value: "com", Usage: "Data center (com, eu, in, com.au, jp, ca, sa, uk, com.cn)"},
					&cli.StringFlag{Name: "scopes", Usage: "Comma-separated OAuth scopes (defaults to all)"},
				},
				Action: func(ctx context.Context, cmd *cli.Command) error {
					return DeviceFlowLogin(
						cmd.String("client-id"),
						cmd.String("client-secret"),
						cmd.String("dc"),
						cmd.String("scopes"),
					)
				},
			},
			{
				Name:  "self-client",
				Usage: "Authenticate via self-client code exchange",
				Flags: []cli.Flag{
					&cli.StringFlag{Name: "code", Required: true, Usage: "Self-client authorization code from Zoho API Console"},
					&cli.StringFlag{Name: "client-id", Required: true, Usage: "Zoho OAuth client ID"},
					&cli.StringFlag{Name: "client-secret", Required: true, Usage: "Zoho OAuth client secret"},
					&cli.StringFlag{Name: "dc", Value: "com", Usage: "Data center"},
					&cli.StringFlag{Name: "server", Usage: "Accounts server URL override"},
				},
				Action: func(ctx context.Context, cmd *cli.Command) error {
					return SelfClientExchange(
						cmd.String("client-id"),
						cmd.String("client-secret"),
						cmd.String("code"),
						cmd.String("dc"),
						cmd.String("server"),
					)
				},
			},
			{
				Name:  "status",
				Usage: "Show current authentication status",
				Action: func(ctx context.Context, cmd *cli.Command) error {
					config, err := ResolveAuth()
					if err != nil {
						return internal.NewAuthError("Not authenticated. Run `zoho auth login` to authenticate.")
					}
					return output.JSON(map[string]any{
						"authenticated": true,
						"source":        config.Source,
						"dc":            config.DC,
						"accounts_url":  config.AccountsURL,
						"token_valid":   config.TokenValid(),
					})
				},
			},
			{
				Name:  "refresh",
				Usage: "Force refresh the access token",
				Action: func(ctx context.Context, cmd *cli.Command) error {
					config, err := ResolveAuth()
					if err != nil {
						return err
					}
					token, err := RefreshAccessToken(config)
					if err != nil {
						return err
					}
					fmt.Fprintln(os.Stderr, "Access token refreshed successfully.")
					truncated := token
					if len(truncated) > 20 {
						truncated = truncated[:20] + "..."
					}
					return output.JSON(map[string]string{"access_token": truncated, "status": "refreshed"})
				},
			},
			{
				Name:  "logout",
				Usage: "Clear stored authentication tokens",
				Action: func(ctx context.Context, cmd *cli.Command) error {
					removed := false
					if _, err := os.Stat(TokensFile()); err == nil {
						os.Remove(TokensFile())
						removed = true
					}
					if _, err := os.Stat(ConfigFile()); err == nil {
						os.Remove(ConfigFile())
						removed = true
					}
					if removed {
						fmt.Fprintln(os.Stderr, "Logged out. Stored credentials removed.")
					} else {
						fmt.Fprintln(os.Stderr, "No stored credentials found.")
					}
					return nil
				},
			},
		},
	}
}
